### ISSUE TITLE
FireHOL fail2ban ipsets issue [help]

### ISSUE THREAD

**Issue by bfranske** [firehol/firehol#272]
I was introduced to FireHOL through the IP Lists site and think it's great so far! Thanks for the hard work! My challenge is that I'm trying to make FireHOL (3.1.1 on Debian Stretch) work in what I think should be a fairly straightforward way but am having a problem actually following the documentation which it seems is not the most up to date. It also doesn't help that ipsets are fairly new to me as well, though very cool.

Essentially I would like to use some of the blacklists (namely firehol_level1 and firehol_level2) in combination with fail2ban for some local blocking of bruteforce attempts. I couldn't actually find documentation on how to apply the firehol_level1 and 2 lists using FireHOL anywhere but I think I got it figured out. I needed to enable each of the lists included inside of those meta-lists idividually with update-ipsets first (is this really right? seems like you should just be able to enable the meta-list and then have it pull the other lists automatically) and then I could modify the firehol.conf file to create the ipsets, to read the netset lists for initial setup, and the blackhole the lists. I've included my config below.

The problem is that the documentation about making fail2ban work seems to be outdated/incorrect for at least some setups. In particular the "postprocess -warn /bin/fail2ban-client reload || return 1" line does not seem to be do anything (though on Debian I have changed it to /usr/bin/fail2ban-client). If I manually run fail2ban-client reload from the command line things seem to work though or if I use the line "postprocess -warn service fail2ban restart || return 1" things seem to work. Unless I've missed it there seems to be no documentation at all about the postprocess command in the FireHOL documentation so I'm unsure of what's going on here and why.

Here's my configuration file which I'm using:
```
version 6

ipv4 ipset create firehol_level1 hash:net
ipv4 ipset addfile firehol_level1 ipsets/firehol_level1.netset
ipv4 ipset create firehol_level2 hash:net
ipv4 ipset addfile firehol_level2 ipsets/firehol_level2.netset

ipv4 blacklist full ipset:firehol_level1 ipset:firehol_level2

interface eth0 internet
        server "http https ssh" accept
        client all accept

postprocess -warn service fail2ban restart || return 1
```

Thanks!

**Comment by bfranske** [comment]
Nope, things are still not working right. I just re-checked my iptables and it looks like at least some of the time the match-set for fail2ban is not being added... If I manually restart fail2ban it seems to show back up...

**Comment by ktsaou** [comment]
Hi,

I think the best way should be to use an ipset and instruct fail2ban to add/remove IPs in this ipset.
However, I am not sure how to do it.

@philwhineray can you help?

**Comment by bfranske** [comment]
Thanks, I do have fail2ban configured to use ipsets. Fail2ban is supposed to automatically create and maintain those ipsets as well as insert an iptables rule to match those sets. The issue seems to be that for some reason when fail2ban is reloaded by the postprocess command in firehol it does not always create the ipsets and/or insert the match rule for them into iptables. It does seem to always do it if I manually run the reload command after firehol is loaded though.

In other words it appears that something about the way firehol invokes the postprocess command prevents fail2ban from working in the same way that it does when it is run without firehol. I'm looking for advice on that or at least some more documentation on the postprocess command which I am unable to find in the official documentation.

**Comment by ktsaou** [comment]
ok, I think I found the problem.

When firehol has fast activation enabled, it runs the post processing commands, before activating the firewall. So, its effects are lost.

This is a bug. Give me a few days to fix it.

**Comment by bfranske** [comment]
Just checking to see if there's a bugfix yet. Thanks!

**Comment by ktsaou** [comment]
Unfortunately, I haven't found the time to fix this. The fix is simple, but it is risky and probably needs some testing.

@philwhineray do you have some time to fix this?

**Comment by philwhineray** [comment]
Really sorry but I don't think I will in the near future.

**Comment by ktsaou** [comment]
ok, I did some research to understand how this works.

We have 2 modes of operation:

1. `FAST_ACTIVATION=0`
2. `FAST_ACTIVATION=1`

When `FAST_ACTIVATION=0`, `postprocess` generates a shell script that includes all the commands to be run.

When `FAST_ACTIVATION=1`, `postprocess` generates the same file, but all the iptables commands are not included there. The iptables statements are restored, after executing the postprocessing file.

So, the diff is this:

In `FAST_ACTIVATION=0`, postprocessing statements are multiplexed with iptables commands. The location of `postprocess` in `firehol.conf` determines at which point it will be executed. This is guaranteed to happen after clearing the old firewall and while the new is activated.

In `FAST_ACTIVATION=1`, postprocessing statements are executed before any iptables commands. The location of `postprocess` in `firehol.conf` is irrelevant. The problem, is that any iptables commands executed by external programs will be lost, since the firehol generated firewall will be restored moments later.

I think the only solution we have is to provide a method for executing commands after the firehol generated firewall has been restored. Something like an exit hook (i.e. run this before you exit). This will guarantee that the command to be executed, will run while the new firewall is active.

How does that sound?

**Comment by bfranske** [comment]
Yes, I think that's how I would expect the postprocess command to work, to be the last thing run before exiting.

**Comment by ktsaou** [comment]
I understand the confusion (I was even confused with this). `postprocess` was meant to be **run during firewall activation** and includes many things that should be run after clearing the old firewall and before the new firewall is activated.

When we introduced `FAST_ACTIVATION=1`, we actually lost the ability to run something *during* firewall activation, since it is now instant. Since it included commands that needed to run before activating the firewall, we ended up needing 2 postprocessing functions:

1. before new firewall activation
2. after new firewall activation

So, I am going to add `postprocess2` that will handle the later (after firewall activation).

**Comment by ktsaou** [comment]
here it is #282

**Comment by nodiscc** [comment]
For anyone looking to make this work on Debian 9 stretch (stable) where firehol still doesn't have this patch (` firehol 3.1.1+ds-1` ), the solution is to edit the fail2ban systemd service so that restarting firehol

 - first stops fail2ban
 - then restarts the actual firehol service
 - then starts fail2ban again.

As you should not edit the unit file in /lib/systemd/system/fail2ban.service, nor completely rewrite it with a completely custom unit file in /etc/systemd/ (as possible future Debian updates of this file would be ignored), the solution is to create an override file. See https://www.freedesktop.org/software/systemd/man/systemd.unit.html and https://www.freedesktop.org/software/systemd/man/systemd.service.html, `Example 2. Overriding vendor settings`:

- Create ` /etc/systemd/system/fail2ban.service.d/local.conf` with the following contents:

```ini
[Service]
# Reset After and PartOf to empty strings
After=
PartOf=
#Define fail2ban as part of the firehol service, make complete firehol startup a prerequisite for starting fail2ban
After=network.target iptables.service firewalld.service firehol.service
PartOf=iptables.service firewalld.service firehol.service
```

- Run `systemctl daemon-reload` to load the unit file override
- Running ` systemctl restart firehol.service` now properly restores fail2ban rules. On boot fail2ban is correctly started _after_ firehol so its rules are not list.

The correct fix to this bug would be submitting a patch to https://github.com/fail2ban/fail2ban/blob/0.11/files/fail2ban.service.in, and clearing https://github.com/firehol/firehol/wiki/fail2ban since the information there is simply outdated, inaccurate, and doesn't work.

I think recommending the use of `postprocess`/`postprocess2` instead of relying on proper, distribution-vetted service management tools is bad practice.

**Comment by nodiscc** [comment]
https://github.com/firehol/firehol/wiki/fail2ban updated
